Tutorial - Nuxt feature flags (#7564)
PostHog/posthog.com: contents/tutorials/nuxt-feature-flags.md

---
title: How to set up feature flags in Nuxt

> date: 2024-01-22

Author: ```suggestion
date: 2024-01-23
```

    api_host: runtimeConfig.public.posthogHost,
  })
  
  return {
    provide: {
      posthog: () => posthogClient
    }
  }
})
```

Once you’ve done this, reload your app. You should begin seeing events in the [PostHog events explorer](https://us.posthog.com/events).

<ProductScreenshot
  imageLight={EventsLight} 
  imageDark={EventsDark} 
  alt="Events in PostHog" 
  classes="rounded"
/>

## Creating a feature flag

With PostHog set up, your app is ready for feature flags. To create one, go to the [feature flags tab](https://us.posthog.com/feature_flags) in PostHog and click "New feature flag." Enter a flag key (like `my-cool-flag`), set the release condition to roll out to 100% of users, and press "Save."

<ProductScreenshot
  imageLight={CreateFlagLight} 
  imageDark={CreateFlagDark} 
  alt="Feature flag created in PostHog" 
  classes="rounded"
/>

You can customize your [release conditions](/docs/feature-flags/creating-feature-flags#release-conditions) with rollout percentages, and [user](/docs/product-analytics/user-properties) or [group properties](/docs/product-analytics/group-analytics) to fit your needs.

> ## Implementing feature flag code on the client side

Author: ```suggestion
## Implementing flag code on the client side
```

Once created, we can add our feature flag to our app. We do this using the [`posthog.onFeatureFlags`](/docs/libraries/js#ensuring-flags-are-loaded-before-usage) callback.

In our example, we fetch our feature flag and update the text on the page based on its value:

```vue file=app.vue
<template>
  <main>
    <h1>{{ title }}</h1>
  </main>
</template>

<script setup>

const title = ref('No variant');
onMounted(() => {
  const { $posthog } = useNuxtApp()
  if ($posthog) {
    const posthog = $posthog()
    if (posthog.isFeatureEnabled('my-cool-flag')) {
      title.value = 'Our flag is enabled!';
    } else {
      title.value = 'Our flag is disabled!';
    }
  }
});

</script>
```

When you run your app now, you should see the updated text.

![New app after adding the flag](../images/tutorials/nuxt-feature-flags/flag-enabled-in-app.png)

> ## Adding PostHog on the server side feature flags

Author: ```suggestion
## Implementing flag code on the server side
```

> You may notice the text flickers when you load your app. This is because PostHog is making a request on the client side to fetch the flag value. To prevent this, there are two alternatives:

Reviewer: ```suggestion
You may notice the text flickers when you load your app. This is because PostHog is making a request on the client side to fetch the flag value. There are two ways to prevent this:
```

1. Fetch the flag on the server side.

> 2. [Bootstrap](/docs/feature-flags/bootstrapping) PostHog on the client side (not covered in this tutorial).

Reviewer: ```suggestion
2. [Bootstrap](/docs/feature-flags/bootstrapping) flags on the client side (not covered in this tutorial).
```

Reviewer: Maybe add a link to the Vue feature flag tutorial? They are similar enough.

```bash
npm install posthog-node
```

Then, we use the PostHog Node library to fetch the feature flag using [`useAsyncData`](https://nuxt.com/docs/api/composables/use-async-data). Replace the code in `app.vue` with the following:

```vue file=app.vue
<template>
  <main>
    <h1>{{ title }}</h1>
  </main>
</template>

<script setup>
import { useAsyncData, useRuntimeConfig } from 'nuxt/app';
import { PostHog } from 'posthog-node';

const { data: titleData, error } = await useAsyncData('titleData', async () => {
  const runtimeConfig = useRuntimeConfig();
  const posthog = new PostHog(
    runtimeConfig.public.posthogPublicKey,
    { host: runtimeConfig.public.posthogHost }
  );

  let returnedValue = '';
  const distinctId = 'placeholder-user-id'
  const isFlagEnabled = await posthog.isFeatureEnabled('my-cool-flag', distinctId);
  if (isFlagEnabled) {
    returnedValue = 'Our flag is enabled!';
  } else {
    returnedValue = 'Our flag is disabled!'
  }
  return returnedValue;
});
const title = computed(() => titleData.value);

</script>
```

Now when you reload your app, the flicker is gone. 

### Setting the correct `distinctId`

> You may have noticed that we set `distinctId = 'placeholder-user-id'` in our flag call above. In production apps, to ensure you fetch the correct flag value for your user, `distinctId` should be set to their unique ID. 

Reviewer: ```suggestion
You may notice that we set `distinctId = 'placeholder-user-id'` in our flag call above. In production apps, to ensure you fetch the correct flag value for your user, `distinctId` should be set to their unique ID. 
```

For logged-in users, you typically use their email as their `distinctId`. However, for logged-out users, you can use the `distinct_id` property from their PostHog cookie:

> ```vue file=app.vue

Reviewer: Can we handle for the situation when the cookie doesn't exist? Right now, it is just blank. No need to add generating an anonymous ID, just an if statement is fine.

Author: Im not sure I follow, there's already an if statement below?

```
if (cookies && cookies.value) {
```

Reviewer: Ahh, maybe change   `let returnedValue = '';` to   `let returnedValue = 'No cookie';` ?

<!-- rest of your code -->

<script setup>
import { useAsyncData, useRuntimeConfig, useCookie } from 'nuxt/app';
import { PostHog } from 'posthog-node';

const { data: titleData, error } = await useAsyncData('titleData', async () => {
  const runtimeConfig = useRuntimeConfig();
  const posthog = new PostHog(
    runtimeConfig.public.posthogPublicKey,
    { host: runtimeConfig.public.posthogHost }
  );

>   let returnedValue = '';

Author: ```suggestion
  let returnedValue = 'No cookie';
```
